File: python/Si3DtoParaview_Processor.py
import os
from Si3DtoParaview import *

root = 'S:/si3D/02_ClearLake/00_HgTests_Trials/'
root = '~/si3D/'
fileSims = ['CL_WQ_06']
concTracer = ['Hg0', 'HgII', 'MeHg']
concTracer = ['DO', 'PON', 'DON', 'NH4', 'NO3', 'POP', 'DOP', 'PO4', 'POC', 'DOC', 'ALG1']

# Code section
for i in fileSims:
    filesim = i
    pathfile = root + i
    pathsave = root + i + '/paraview'

    # To obtain simulation information
    os.chdir(pathfile)

    fid = open('si3d_inp.txt', 'r')
    lines = fid.readlines()
    year = lines[5]
    month = lines[6]
    day = lines[7]
    hour = lines[8]
    year = str(year.split()[2])
    month = str(month.split()[2])
    day = str(day.split()[2])
    hour = str(hour.split()[2])
    minute = str(hour[2:4])
    hour = str(hour[0:2])
    startDate = year + '-' + month + '-' + day + ' ' + hour + ':' + minute + ':00'

    # To obtain dx
    dx = lines[16]
    dx = str(dx)
    dx = float(dx[14:34])

    # To obtain dz
    dz = lines[18]
    dz = str(dz)
    dz = float(dz[14:34])

    # To obtain dt
    dt = lines[22]
    dt = str(dt)
    dt = float(dt[14:34])

    # To obtain idz
    idz = lines[23]
    idz = str(idz)
    idz = int(idz[14:34])
    if idz == 0:
        deltaZ = 'constant'
    elif idz == -1:
        deltaZ = 'variable'
    # To obtain iTurb
    iTurb = lines[75]
    iTurb = str(iTurb)
    iTurb = int(iTurb[14:34])
    # To obtain output settings
    itspf = lines[74]
    itspf = str(itspf)
    itspf = int(itspf[14:34])
    itspfh = lines[63]
    itspfh = str(itspfh)
    itspfh = int(itspfh[14:34])
    itspftr = lines[104]
    itspftr = str(itspftr)
    itspftr = int(itspftr[14:34])

    nTracer = lines[101]
    nTracer = str(nTracer)
    nTracer = int(nTracer[14:34])

    if (itspf != itspfh) or (itspf != itspftr) or (itspfh != itspftr):
        raise ValueError('Error the variables itspf and itspfh must be the same. If tracers are modeled, itspftr, itspf, and itspfh must be the same')
    if nTracer > 0:
        if nTracer != len(concTracer):
            raise ValueError('Number of tracers modeled different from number of units provided as input')

    if os.path.isfile('si3d_3D'):
        n_frames = Si3DtoParaview(pathfile, pathsave, startDate, deltaZ, dx, dz, dt, iTurb, itspf, nTracer, concTracer)
    else:
        raise ValueError('3D File not Found')

File: matlablibrary/Si3DtoParaview_Processor.m
% NEED TO PUT TEXT TO EXPLAIN THE PROCESSOR.

clc;
close all;
clear variables;

% root = 'S:\si3D\Alicia_Runs\';
% FileSim = 'si3dTest';
root = 'S:\si3D\02_ClearLake\00_HgTests\';
% FileSims = {'CL_run24p_hydro'};
FileSims = {'002_CL'};
concTracer = {'DO', 'PON', 'DON', 'NH4', 'NO3', 'POP', 'DOP', 'PO4', 'POC', 'DOC', 'ALG1', 'SS1', 'Hg0', 'HgII', 'MeHg'};

%% Code section
for i = 1:length(FileSims)
    FileSim = char(FileSims(i));
    PathFile = [root,FileSim];
    PathSave = [root,FileSim,'\Paraview'];

    % To obtain simulation information like dt,dz,dx and date
    cd(PathFile)
    inputFile = readlines('si3d_inp.txt');
    year = inputFile(6);
    month = inputFile(7);
    day = inputFile(8);
    hour = inputFile(9);
    year = char(regexp(year,'\d+\.?\d*|-\d+\.?\d*|\.?\d+|-\.?\d+','match'));
    month = char(regexp(month,'\d+\.?\d*|-\d+\.?\d*|\.?\d+|-\.?\d+','match'));
    day = char(regexp(day,'\d+\.?\d*|-\d+\.?\d*|\.?\d+|-\.?\d+','match'));
    hour = char(regexp(hour,'\d+\.?\d*|-\d+\.?\d*|\.?\d+|-\.?\d+','match'));
    minute = hour(3:4);
    hour = hour(1:2);
    StartDate = [year,'-',month,'-',day,' ',hour,':',minute,':00'];
    
    % Obtain dx from si3d_inp.txt
    dx = inputFile(17);
    dx = char(dx);
    dx = str2double(dx(15:34));
    % Obtain dz from si3d_inp.txt
    dz = inputFile(19);
    dz = char(dz);
    dz = str2double(dz(15:34));
    % Obtain dt from si3d_inp.txt
    dt = inputFile(23);
    dt = char(dt);
    dt = str2double(dt(15:34));
    
    % Obtain idz from si3d_inp.txt and know if si3d_layer.txt is necessary
    idz = inputFile(24);
    idz = char(idz);
    idz = idz(15:34);
    idz = str2double(idz);
    if idz == 0
        DeltaZ = 'constant';
    elseif idz == -1
        DeltaZ = 'variable';
    end
    % Obtain iTurb from si3d_inp.txt. This indicates whether ptrack file
    % has turbulent parameters
    iTurb = inputFile(76);
    iTurb = char(iTurb);
    iTurb = str2double(iTurb(15:34));
    % Obtain itspf from si3d_inp.txt. This indicates whether time steps
    % saved started from the start of the simulation or after a certain
    % time step
    itspf = inputFile(75);
    itspf = char(itspf);
    itspf = str2double(itspf(15:34));

    itspfh = inputFile(64);
    itspfh = char(itspfh);
    itspfh = itspfh(15:34);
    itspfh = str2double(itspfh);
    itspftr = inputFile(105);
    itspftr = char(itspftr);
    itspftr = itspftr(15:34);
    itspftr = str2double(itspftr);

    nTracer = inputFile(102);
    nTracer = char(nTracer);
    nTracer = nTracer(15:34);
    nTracer = str2double(nTracer);

    if nTracer > 0
        if nTracer ~= length(concTracer)
            error('Number of tracers modeled different from number of units provided as input')
        end
        if itspf ~= itspfh || itspf ~= itspftr || itspfh ~= itspftr
            error('If tracers are modeled, itspftr, itspf, and itspfh must be the same')
        end
    else
        if itspf ~= itspfh
            error('Error the variables itspf and itspfh must be the same')
        end
    end
    
    %% To generate Paraview files
    if isfile('si3d_3D')
        % File exists.
        n_frames = Si3DtoParaview(PathFile,PathSave,StartDate,DeltaZ,dx,dz,dt,iTurb,itspf,nTracer,concTracer);
    else
        disp('Error 3D file was not found')
    end
end